src/ppx/ppx_eliom_utils.ml

```diff
@@ -461,8 +469,277 @@ end
      ]
    ]
 *)
-module Shared = struct
+module Rpc = struct
+  type error = No_arguments | Missing_argument_type | Fatal_error
+
+  let print_error ~loc (e : error) =
+    let string_of_error =
+      match e with
+      | No_arguments ->
+          "No arguments"
+      | Missing_argument_type ->
+          "Missing argument type, (argument:type) format expected"
+      | Fatal_error ->
+          "Fatal error"
+    in
+    Location.raise_errorf ~loc "%s" string_of_error
+
+  let function_name fun_name_pattern =
+    let loc = fun_name_pattern.ppat_loc in
+    match fun_name_pattern with
+    | [%pat? [%p? {ppat_desc= Ppat_var ident}]] ->
+        ident.txt
+    | _ ->
+        print_error ~loc Fatal_error
+
+  let rpc_name fun_name_pattern =
+    let filename =
+      Filename.(!Location.input_name |> chop_extension |> basename)
+    in
+    Format.sprintf "%s.%s" filename (function_name fun_name_pattern)
+
+  let pat_to_exp pattern =
+    match pattern with
+    | {ppat_desc= Ppat_var ident; _} ->
+        Exp.ident (Location.mkloc (Longident.parse ident.txt) pattern.ppat_loc)
+    | {ppat_desc= Ppat_construct _; _} ->
+        AC.unit ()
+    | _ ->
+        print_error ~loc:pattern.ppat_loc Fatal_error
+
+  let apply_function_expr ~loc fun_name args_list =
+    Exp.apply
+      (Exp.ident (Location.mkloc (Longident.parse fun_name) loc))
+      args_list
+
+  let rec args_parser expr arg_list =
+    match expr with
+    | [%expr fun ([%p? pattern] : [%t? typ]) -> [%e? expr']] ->
+        args_parser expr' ((pattern, typ, Asttypes.Nolabel) :: arg_list)
+    | [%expr fun () -> [%e? expr']] ->
+        args_parser expr'
+          ((AC.punit (), AC.tconstr "unit" [], Asttypes.Nolabel) :: arg_list)
+    | { pexp_desc=
```

Comment:
```ocaml
| { pexp_desc=
￼          Pexp_fun
￼            ( arg
￼            , None
￼            , {ppat_desc= Ppat_constraint (pattern, typ)}
￼            , expr' ) } 
  ->
￼        args_parser expr' ((pattern, typ, arg) :: arg_list)
```
I think you could share a bit more code between labelled and optionnal args by doing something like that. But not sure what are the other enumerated type cases.

Reply:
It seems to be only `Nolabel` and already matched before, so I think what I proposed you should be working.